go/cmd/vtbackup/cli/vtbackup.go

```diff
@@ -229,7 +229,14 @@ func run(_ *cobra.Command, args []string) error {
 		<-ctx.Done()
 	}()
 
+	// Some plugins use OnRun to initialize. Use a channel to signal to
+	// ourselves that these plugins have been initialized.
+	runCh := make(chan struct{})
+	servenv.OnRun(func() {
+		close(runCh)
+	})
 	go servenv.RunDefault()
+	<-runCh
```

Comment:
Thanks for the suggestion, I'll go with the middle option for now (`opentsdb.AwaitInit`).

Reply:
Actually going with second option since we don't know if user will select `--stats_backend opentsdb` or something else.

Reply:
@ajm188 ah that's true 😳 

Do you prefer I switch to that or are you OK with what I pushed up for `stats.AwaitBackend`?

Reply:
so i like yours _structurally_ (since it's possible of course to compile vitess without a particular plugin\_ file), however, the problem is [this](https://github.com/vitessio/vitess/pull/14113/files/bc2178aa134f2bb08d7f77c3ddcabf5385c04f8a..83932bc282beec1d91bf11d74ceea8e1549e8db1#diff-1d61b5058809d9d6f28e6a6cd22b596a68220afcfad4a84334e8d962d63b9372R236) close happens during `init()` and not during `servenv.OnRun`

we need some mechanism for a plugin to signal back to the main `stats` package that it's been fully initialized.

Reply:
something maybe like this:

```
diff --git a/go/stats/export.go b/go/stats/export.go
index 8bda85c87b..07381f9929 100644
--- a/go/stats/export.go
+++ b/go/stats/export.go
@@ -29,6 +29,7 @@ package stats
 
 import (
 	"bytes"
+	"context"
 	"expvar"
 	"fmt"
 	"strconv"
@@ -218,6 +219,12 @@ func RegisterPushBackend(name string, backend PushBackend) {
 		log.Fatalf("PushBackend %s already exists; can't register the same name multiple times", name)
 	}
 	pushBackends[name] = backend
+	backends[name] = &struct {
+		once sync.Once
+		ch   chan struct{}
+	}{
+		ch: make(chan struct{}),
+	}
 	if emitStats {
 		// Start a single goroutine to emit stats periodically
 		once.Do(func() {
@@ -226,6 +233,36 @@ func RegisterPushBackend(name string, backend PushBackend) {
 	}
 }
 
+var (
+	backends = map[string]*struct {
+		once sync.Once
+		ch   chan struct{}
+	}{}
+)
+
+func InitializeBackend(backend string) {
+	if d, ok := backends[backend]; ok {
+		d.once.Do(func() { close(d.ch) })
+	}
+}
+
+func AwaitBackend(ctx context.Context) error {
+	pushBackendsLock.Lock()
+	defer pushBackendsLock.Unlock()
+
+	d, ok := backends[statsBackend]
+	if !ok {
+		return fmt.Errorf("...")
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-d.ch:
+		return nil
+	}
+}
+
 // emitToBackend does a periodic emit to the selected PushBackend. If a push fails,
 // it will be logged as a warning (but things will otherwise proceed as normal).
 func emitToBackend(emitPeriod *time.Duration) {
diff --git a/go/stats/opentsdb/init.go b/go/stats/opentsdb/init.go
index 51186ad765..462a442501 100644
--- a/go/stats/opentsdb/init.go
+++ b/go/stats/opentsdb/init.go
@@ -41,6 +41,8 @@ func Init(prefix string) {
 		if err != nil {
 			log.Infof("Failed to initialize singleton opentsdb backend: %v", err)
 		} else {
+			defer stats.Initialize("opentsdb")
+
 			singletonBackend = backend
 			log.Info("Initialized opentsdb backend.")
 		}
```

Reply:
@ajm188 and I chatted about this part:

> the problem is [this](https://github.com/vitessio/vitess/pull/14113/files/bc2178aa134f2bb08d7f77c3ddcabf5385c04f8a..83932bc282beec1d91bf11d74ceea8e1549e8db1#diff-1d61b5058809d9d6f28e6a6cd22b596a68220afcfad4a84334e8d962d63b9372R236) close happens during init() and not during servenv.OnRun

It turns out `opentsdb` is a bit special in this regard in that it exposes a `Init` function that uses `servenv.OnRun` under the hood. So the full sequence is:

 1. https://github.com/vitessio/vitess/blob/83932bc282beec1d91bf11d74ceea8e1549e8db1/go/cmd/vtbackup/cli/plugin_opentsdb.go#L24
 2. https://github.com/vitessio/vitess/blob/83932bc282beec1d91bf11d74ceea8e1549e8db1/go/stats/opentsdb/init.go#L36-L38
 3. https://github.com/vitessio/vitess/blob/83932bc282beec1d91bf11d74ceea8e1549e8db1/go/vt/servenv/run.go#L40-L43
 4. https://github.com/vitessio/vitess/blob/83932bc282beec1d91bf11d74ceea8e1549e8db1/go/stats/opentsdb/init.go#L51-L58
 5. https://github.com/vitessio/vitess/blob/83932bc282beec1d91bf11d74ceea8e1549e8db1/go/stats/export.go#L228-L236

It's probably worth making the `opentsdb` package and consumer plugins more like the rest of the code base at some point, but will leave things as they are for now if that's alright.

Reply:
yep, you're totally right! this is looking good to me now